---
 fs/jbd2/journal.c   |    1 +
 security/security.c |    1 +
 2 files changed, 2 insertions(+)

--- a/security/security.c
+++ b/security/security.c
@@ -459,6 +459,7 @@ int security_inode_unlink(struct inode *
 		return 0;
 	return security_ops->inode_unlink(dir, dentry);
 }
+EXPORT_SYMBOL(security_inode_unlink);

 int security_inode_symlink(struct inode *dir, struct dentry *dentry,
 			    const char *old_name)
